lfs/util.go

```diff
@@ -310,3 +310,35 @@ func FileExistsOfSize(path string, sz int64) bool {
 
 	return !fi.IsDir() && fi.Size() == sz
 }
+
+func CopyFileContents(src string, dst string) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+	_, err = io.Copy(out, in)
+	if err != nil {
+		return err
+	}
+	err = out.Sync()
+	return err
+}
+
+func LinkOrCopy(src string, dst string) error {
+	err := os.Link(src, dst)
```

Comment:
I would make this safer by doing 2 things:

1. Validate using `os.SameFile` to make sure `src` and `dst` are not the same
2. Instead of calling `CopyFileContents` with `dst` as the direct output, instead copy to a temp file and then `os.Rename` on success. This will be more robust under unusual cases like forced termination.